Initial updates

marohrdanz opened the pull request:
The key changes in this pull request are:

- Update the communication with NGCHM to get correct color map
- Added a 'demo' page the loads demo data (demo_data directory)
   - as with the 2D scatter plot, it's nice to have a simple page of test data for development of new features
- Added option for orientation axes at the origin (related to: #2)
  - These axes don't move with the center of rotation but instead remain fixed at the origin. 
     (If the user shifts the center of the scene, the center of rotation shifts.)
- Added option to color axes (#3)
- Updated hover over point to show point name and coordiantes in upper right corner,  
   and pass message to other plugins (#5)
- A few other minor cosmetic changes (these are mentioned in the commit messages)

marohrdanz committed:
Remove 'group' from Vanodi registration

Removed the 'group' field from the vanodi registration--this 'group'
is probably a left-over entry from an old version of the API.

marohrdanz committed:
Add initial demo plot

Added initial demo plot: simple HTML page (demo-plot.html),
script to load demo data (demo-plot.js),
and a directory of a script for creating & storing demo data
(/demo_data).

These are for development and testing only (it's easier to add
new features and testing existing ones w/o the added
complication of the NGCHM)

marohrdanz committed:
Add axis labels

marohrdanz committed:
Add initial interface with NGCHM

Added the initial interface to the NGCHM (with vanodi, in ngchm.js).
This is still a bit of a work in progress, but the initial features
(plotting data with correct colormap) are present.

Also renamed the 'demo-plot.js' -> scatterPlot3D.js

marohrdanz committed:
Add feature for picking points under cursor

Followed the model from this post
https://threejsfundamentals.org/threejs/lessons/threejs-picking.html
to create functionality for picking points under the mouse.

At the moment, the id of the point under the mouse is simply
shown in the span with id = tooltip on the top of the page.
Also, the color of the whole batch containing the selected
point is changed to red. The plan is for this to evolve into
a local tooltip over the point, only highlighting the
selected point, and posting a message to the NGCHM to highlight
the point on the NGCHM, etc. etc.

marohrdanz committed:
Remove superseded scripts

The functionality of plot3d.js is now in scatterPlot3D.js, and the
functionality of scatter3d.js is now in ngchm.js (roughly speaking)

marohrdanz committed:
Move D3 script to 'demo_data' directory

Moved script to avoid confusion: this d3 script is only used to
load the demo data (D3 has an easy function to load data asynchronously).

marohrdanz committed:
Create separate function for axes drawing

Refactor code to put drawing of axes in a separate function.

marohrdanz committed:
Ensure date read in as numeric and not string

marohrdanz committed:
Add option for 'box' or 'origin' coordinates

Most people prefer the original box-style coordinate rendering.
However some prefer the coordinates at the origin in a more
tranditional Cartesian rendering. This commit adds the option in the
gear menu for the user to choose their preference.

Also adds an option to choose the point size. (this will eventually
become a slider)

marohrdanz committed:
Fix rendering bug: previous choices showup

I *think* maybe this fixes a bug we were having: if new plot options
were chosen (such as a different color map), sometimes when moving
around the previous choices would be drawn.

We moved away from animate with requestAnimationFrame because (on my
laptop at least) it causes the computer fan to come on. So I think
the bug was caused by using old-rendered versions of the scene.
(I'm open to being wrong about this...so we'll have to keep an eye
out for this bug).

marohrdanz committed:
Clarify help text for Axes Render Type

marohrdanz committed:
Refactor to move functions & demo page xyz coords

Did a gentle refactor that moved some larger functions out of the
main exported createPlot function. Also added a function to show
the XYZ coordinates as a method in the PickHelper class.

marohrdanz committed:
Add method to emphasize highlighted point

Added two methods to PickHelper:

 - PickHelper.highlightPoint adds a transparent sphere around the
   highlighted point
 - PickHelper.clearHighlightedPoints removes that sphere

These are called as the mouse moves around, adding and removing
the highlights as the mouse moves.

marohrdanz committed:
Add option to change highlight color

Added an option for the user to specify the highlight color, and
made the opacity a little less.

marohrdanz committed:
Fix render bugs & highlighting point under cursor

Fixed a rendering bug: when the user would change plot options (e.g.
point color) sometimes the scene would revert to the old version as
the user was manipulating the scene. This was caused by not clearing
out the old scene before redrawing.

Also fixed the highlighting: the initial version in this branch was
changing the color of the whole group red (this was just to test things
out, and was never intended as the final version). This commit uses
the transparent sphere to highlight the data point under the cursor.
This information is then sent to the NGCHM to highlight the point
on the heatmap.

marohrdanz committed:
Refactor for maintainability; fix selection bug

Working on getting the highlighting (when user hovers over point)
and selections (when user selects points on NGCHM) to work properly.
This commit:

- fixes some bugs with selections
- adds a temporary library and bit of code to monitor performance
  of WebGL which will be removed in future commits
  - see:
   http://learningthreejs.com/blog/2013/06/25/monitor-rendering-performance-within-threejs/
- does some refactoring to split things into several files to
  improve comprehensibility (there is still more work to be done here)

This commit is still something of a WIP.

marohrdanz committed:
Put renderer performance testing in demo-plot only

Moved the renderer performance testing code to only the demo-plot
page. Also removed the animate-every-frame from the main index.html
page so that the main page only renders when the scene changes.
Those renders are basically manual: whenever anything changes,
there is a call to the renderer.

marohrdanz committed:
Refactor for maintainability

Moved the functions for drawing axes to their own file.

marohrdanz committed:
Make minor cosmetic changes to demo page

Mostly just changing the size of the WebGL profiler box.

marohrdanz committed:
Refactor to add function for creating scales

Moved the code for creating the sX, sY, sZ to its own function.

Set the axis length parameter to 9. (Still haven't found a good way
to not hard-code this parameter).

marohrdanz committed:
Reduce axis label font & add 'tooltip'

Reduced the font size for the axis labels.

Added the point name and XYZ coordiantes of the hovered-over
point to the top of the index page (this was already present on
the demo page)

marohrdanz committed:
Add extra padding between points & axes

marohrdanz committed:
Add option to make axes colored

Added an option in the Gear Menu to allow for the three axes to
be a different color.

marohrdanz committed:
Add option for axes at origin (compass rose)

Removed the 'addOriginAxes' because no one liked it. Removed both the
function to create them, and the option from the Gear Menu.

These origin axes here have been essentially replaced by an option for
displaying axes at the origin that act as a compass rose. These are
shown in addition to the box axes, rather than as an alternative.

The origin axes (that have been removed) were drawn as lines. These
new 'compass rose' axes are instead drawn as cylinders and cones
because this gives more flexibility in choosing widths, textures, etc.

bmbroom reviewed (approved):
Looking like a good point for a merge.

marohrdanz committed:
Make selection more opaque  & allow color off rose

Made the selected points material more opaque because it was difficult
to see which points where selected when choosing points from the
NGCHM.

Also updated the color specification of the rose axes to allow user
so that these axes will be gray if the user turned off the color
axes option.

marohrdanz commented:
That force push was because I rebased to remove a WIP commit that shouldn't have been pushed here :see_no_evil:
